File: portfolio-api/src/routes/exchanges.ts
import { Elysia } from "elysia";
import { listExchanges, createExchange, deleteExchange } from "../db/queries/exchanges";
import type { CreateExchangeRequest } from "../types/api";

export const exchangePlugin = new Elysia({ prefix: "/api/exchanges" })
  .get("/", () => ({ exchanges: listExchanges() }))
  .post("/", ({ body, set }) => {
    const data = body as CreateExchangeRequest;
    if (!data.name || !data.type) { set.status = 400; return { error: "name and type are required" }; }
    const exchange = createExchange(data.name, data.type);
    set.status = 201;
    return { exchange };
  })
  .delete("/:id", ({ params, set }) => {
    const id = parseInt(params.id);
    if (isNaN(id)) { set.status = 400; return { error: "Invalid id" }; }
    deleteExchange(id);
    return { ok: true };
  });

File: fastapi/app/api/routes/exchanges.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session

from app import crud
from app.core.db import get_session
from app.models import ExchangeCreate

router = APIRouter(prefix="/exchanges", tags=["exchanges"])


@router.get("/")
def list_exchanges(session: Session = Depends(get_session)) -> dict:
    return {"exchanges": crud.list_exchanges(session)}


@router.post("/", status_code=201)
def create_exchange(body: ExchangeCreate, session: Session = Depends(get_session)) -> dict:
    exchange = crud.create_exchange(session, body)
    return {"exchange": exchange}


@router.delete("/{exchange_id}")
def delete_exchange(exchange_id: int, session: Session = Depends(get_session)) -> dict:
    exchange = crud.get_exchange(session, exchange_id)
    if not exchange:
        raise HTTPException(status_code=404, detail="Exchange not found")

    # An asset still pointing here would fail the foreign key and surface as a 500 with
    # nothing the user can act on. Name the assets instead so they know what to move.
    blocking = crud.list_assets_by_exchange(session, exchange_id)
    if blocking:
        symbols = ", ".join(a.symbol for a in blocking)
        raise HTTPException(
            status_code=409,
            detail=(
                f"{exchange.name} still holds {len(blocking)} "
                f"position{'s' if len(blocking) != 1 else ''} ({symbols}). "
                "Delete or move them first."
            ),
        )

    crud.delete_exchange(session, exchange)
    return {"ok": True}
